Project/Source/Navigation/NavMesh.cpp

```diff
@@ -130,297 +505,133 @@ bool NavMesh::Build() {
-		for (int i = 0; i < pmesh->npolys; ++i) {
-			if (pmesh->areas[i] == RC_WALKABLE_AREA)
-				pmesh->areas[i] = SAMPLE_POLYAREA_GROUND;
-
-			if (pmesh->areas[i] == SAMPLE_POLYAREA_GROUND || pmesh->areas[i] == SAMPLE_POLYAREA_GRASS || pmesh->areas[i] == SAMPLE_POLYAREA_ROAD) {
-				pmesh->flags[i] = SAMPLE_POLYFLAGS_WALK;
-			} else if (pmesh->areas[i] == SAMPLE_POLYAREA_WATER) {
-				pmesh->flags[i] = SAMPLE_POLYFLAGS_SWIM;
-			} else if (pmesh->areas[i] == SAMPLE_POLYAREA_DOOR) {
-				pmesh->flags[i] = SAMPLE_POLYFLAGS_WALK | SAMPLE_POLYFLAGS_DOOR;
+	for (int y = 0; y < th; ++y) {
+		for (int x = 0; x < tw; ++x) {
+			TileCacheData tiles[MAX_LAYERS];
+			memset(tiles, 0, sizeof(tiles));
+			int ntiles = rasterizeTileLayers(&verts[0], nverts, ntris, ctx, chunkyMesh, x, y, cfg, tiles, MAX_LAYERS);
+
+			for (int i = 0; i < ntiles; ++i) {
+				TileCacheData* tile = &tiles[i];
+				status = tileCache->addTile(tile->data, tile->dataSize, DT_COMPRESSEDTILE_FREE_DATA, 0);
+				if (dtStatusFailed(status)) {
+					dtFree(tile->data);
+					tile->data = 0;
+					continue;
+				}
 			}
 		}
+	}
 
-		dtNavMeshCreateParams params;
-		memset(&params, 0, sizeof(params));
-		params.verts = pmesh->verts;
-		params.vertCount = pmesh->nverts;
-		params.polys = pmesh->polys;
-		params.polyAreas = pmesh->areas;
-		params.polyFlags = pmesh->flags;
-		params.polyCount = pmesh->npolys;
-		params.nvp = pmesh->nvp;
-		params.detailMeshes = dmesh->meshes;
-		params.detailVerts = dmesh->verts;
-		params.detailVertsCount = dmesh->nverts;
-		params.detailTris = dmesh->tris;
-		params.detailTriCount = dmesh->ntris;
-		params.walkableHeight = agentHeight;
-		params.walkableRadius = agentRadius;
-		params.walkableClimb = agentMaxClimb;
-		rcVcopy(params.bmin, pmesh->bmin);
-		rcVcopy(params.bmax, pmesh->bmax);
-		params.cs = cfg.cs;
-		params.ch = cfg.ch;
-		params.buildBvTree = true;
-
-		if (!dtCreateNavMeshData(&params, &navData, &navDataSize)) {
-			LOG("Could not build Detour navmesh.");
-			return false;
-		}
-
-		navMesh = dtAllocNavMesh();
-		if (!navMesh) {
-			dtFree(navData);
-			navData = nullptr;
-			navDataSize = 0;
-			LOG("Could not create Detour navmesh");
-			return false;
-		}
-
-		dtStatus status;
-
-		status = navMesh->init(navData, navDataSize, DT_TILE_FREE_DATA);
-		if (dtStatusFailed(status)) {
-			dtFree(navData);
-			navData = nullptr;
-			navDataSize = 0;
-			LOG("Could not init Detour navmesh");
-			return false;
-		}
+	// Build initial meshes
+	for (int y = 0; y < th; ++y)
+		for (int x = 0; x < tw; ++x)
+			tileCache->buildNavMeshTilesAt(x, y, navMesh);
+	ctx->stopTimer(RC_TIMER_TOTAL);
 
-		status = navQuery->init(navMesh, 2048);
-		if (dtStatusFailed(status)) {
-			LOG("Could not init Detour navmesh query");
-			return false;
-		}
+	const dtNavMesh* nav = navMesh;
+	int navmeshMemUsage = 0;
+	for (int i = 0; i < nav->getMaxTiles(); ++i) {
+		const dtMeshTile* tile = nav->getTile(i);
+		if (tile->header)
+			navmeshMemUsage += tile->dataSize;
```

Comment:
This if should be either collapsed into one line or braces should be added. There are a few more instances in this file.